docs/spec/spec-proposals/f1-fee-distribution/f1_fee_distr.tex

```diff
@@ -85,36 +105,61 @@ \subsection{Slashing}
+Note that this inflation model is more fair than the currently implemented inflation algorithm, as this has inflation reward delegators not just validators.
 
 \subsection{Delegation updates}
 Updating your delegation amount is equivalent to withdrawing earned rewards and a fully independent new delegation occuring in the same block.
 
 \subsection{Jailing / being kicked out of the validator set}
-This basically requires no change. In each block you only iterate over the currently bonded validators. So you simply don't update the "total accrued fees this period" variable for jailed / non-bonded validators. Withdrawing requires \textit{no} special casing here!
+This basically requires no change.
+In each block you only iterate over the currently bonded validators.
+So you simply don't update the "total accrued fees this period" variable for jailed / non-bonded validators.
+Withdrawing requires \textit{no} special casing here!
 
 \section{State pruning}
 You will notice that in the main scheme there was no note for pruning entries from state.
 We can in fact prune quite effectively.
 Suppose for the sake of exposition that there is at most one delegation / withdrawal to a particular validator in any given block.
 Then each delegation is responsible for one addition to state.
 Only the next period, and this delegator's withdrawal could depend on this entry. Thus once this delegator withdraws, this state entry can be pruned.
-For the entry created by the delegator's withdrawal, that is only required by the creation of the next period. Thus once the next period is created, that withdrawal's period can be deleted.
+For the entry created by the delegator's withdrawal, that is only required by the creation of the next period.
+Thus once the next period is created, that withdrawal's period can be deleted.
 
-This can be easily adapted to the case where there are multiple delegations / withdrawals per block. Keep a counter per state entry for how many delegations need to be cleared. (So 1 for each delegation in that block which created that period, 0 for each withdrawal) When creating a new period, check that the previous period (which had to be read anyway) doesn't have a count of 0. If it does have a count of 0, delete it. When withdrawing, decrement the period which created this delegation's counter by 1. If that counter is now 0, delete that period.
+This can be easily adapted to the case where there are multiple delegations / withdrawals per block.
+Keep a counter per state entry for how many delegations need to be cleared.
+(So 1 for each delegation in that block which created that period, 0 for each withdrawal)
+When creating a new period, check that the previous period (which had to be read anyway) doesn't have a count of 0.
+If it does have a count of 0, delete it.
+When withdrawing, decrement the period which created this delegation's counter by 1.
+If that counter is now 0, delete that period.
```

Comment:
oh look, you've invented reference counting!

I wonder if we can do this in general actually... would be interesting

Reply:
I also wonder if we could do it in a way which wouldn't break the Merkle root - e.g. if a validator can prove that an entry will never be accessed, they can keep the IAVL key but delete the value. Then these upgrades could be non-breaking.

Reply:
LOL, you're right is just reference counting!

The issue with this pruning is that any new syncing node will have to persist this state forever. (As they'd wouldn't get the proof of it never be accessed) Or if a node synced after it was put into state, they will have to keep that entry forever. This makes me feel like it ought to be in consensus, but it could be more intrinsic within the merkle tree rather than the leaf value.